### ISSUE TITLE
LSP folding request results in java.lang.IllegalArgumentException

### ISSUE THREAD

**Issue by agonzalezgaliana** [eclipse/xtext-core#1778]
Currently testing the latest 2.26.0.M2 version I came up with a java.lang.IllegalArgumentException when a
`{jsonrpc: '2.0', id: 30, method: 'textDocument/foldingRange', params: {…}}` request is received. The stack trace of the error is the following:

Nov 12, 2021 9:29:31 AM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
SEVERE: Internal error: java.lang.IllegalArgumentException: length -1 is < 0
java.util.concurrent.CompletionException: java.lang.IllegalArgumentException: length -1 is < 0
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:346)
	at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:704)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088)
	at org.eclipse.xtext.ide.server.concurrent.AbstractRequest.logAndCompleteExceptionally(AbstractRequest.java:73)
	at org.eclipse.xtext.ide.server.concurrent.ReadRequest.lambda$doRun$0(ReadRequest.java:69)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalArgumentException: length -1 is < 0
	at org.eclipse.xtext.util.TextRegion.<init>(TextRegion.java:21)
	at org.eclipse.xtext.ide.editor.folding.DefaultFoldingRangeProvider.buildSignificantRegion(DefaultFoldingRangeProvider.java:130)
	at org.eclipse.xtext.ide.editor.folding.DefaultFoldingRangeProvider.acceptObjectFolding(DefaultFoldingRangeProvider.java:98)
	at org.eclipse.xtext.ide.editor.folding.DefaultFoldingRangeProvider.computeObjectFolding(DefaultFoldingRangeProvider.java:83)
	at org.eclipse.xtext.ide.editor.folding.DefaultFoldingRangeProvider.getFoldingRanges(DefaultFoldingRangeProvider.java:60)
	at org.eclipse.xtext.ide.server.folding.FoldingRangeService.createFoldingRanges(FoldingRangeService.java:37)
	at org.eclipse.xtext.ide.server.LanguageServerImpl.lambda$foldingRange$51(LanguageServerImpl.java:1039)
	at org.eclipse.xtext.ide.server.WorkspaceManager.doRead(WorkspaceManager.java:438)
	at org.eclipse.xtext.ide.server.LanguageServerImpl.foldingRange(LanguageServerImpl.java:1038)
	at org.eclipse.xtext.ide.server.LanguageServerImpl.lambda$foldingRange$50(LanguageServerImpl.java:1026)
	at org.eclipse.xtext.ide.server.concurrent.ReadRequest.lambda$doRun$0(ReadRequest.java:66)
	... 5 more

This occurs in an Xtext project that defines a "complex" DSL, however, the error does not appear with a different project that uses the grammar defined in one of the [Xtext tutorials](https://www.eclipse.org/Xtext/documentation/102_domainmodelwalkthrough.html)

Not sure if the problem is in the new class `org.eclipse.xtext.ide.editor.folding.DefaultFoldingRangeProvider` line 130:

`return new TextRegion(offset, index - offset - 1);`

If I see it right, the `TextRegion` will be always < 0 in the first iteration because `index = offset`

Perhaps the problem is in my grammar? Maybe some grammars simply do not allow correct folding?

Some help here would be appreciated.

**Comment by cdietrich** [comment]
this might be a bug. can you please provide a minimal grammar + unit test?
ypu might also debug why DefaultFoldingRangeProvider.buildSignificantRegion creates a region with -1
i assume
if (lineInfo.getLine() != startLine) {
should guard against this, but does not for you

**Comment by cdietrich** [comment]
ping @agonzalezgaliana

**Comment by lars-reimann** [comment]
I'm also having this issue with `2.26.0.M2` while typing when the DSL program has syntax errors. The same grammar worked in earlier versions. I'll see whether I can create a minimal test case.

**Comment by cdietrich** [comment]
```
LineAndColumn lineInfo = NodeModelUtils.getLineAndColumn(node, index);
```
returns
```
LineAndColumn [
  line = 4
  column = 1
]
```
whilst startLine = 3

**Comment by cdietrich** [comment]
org.eclipse.xtext.util.ITextRegionWithLineInformation.getLineNumber() seems to be zero based.

**Comment by cdietrich** [comment]
=> pot. fix

```
	protected ITextRegion buildSignificantRegion(ITextRegion significantRegion, INode node) {
		if (significantRegion == null || node == null) {
			return null;
		}
		int offset = significantRegion.getOffset();
		int endOffset = significantRegion.getOffset() + significantRegion.getLength();

		int startLine;
		int endLine;
		if (significantRegion instanceof ITextRegionWithLineInformation) {
			ITextRegionWithLineInformation lineInfoRegion = (ITextRegionWithLineInformation) significantRegion;
			startLine = lineInfoRegion.getLineNumber() + 1;
			endLine = lineInfoRegion.getEndLineNumber() + 1;
		} else {
			startLine = NodeModelUtils.getLineAndColumn(node, offset).getLine();
			endLine = NodeModelUtils.getLineAndColumn(node, endOffset).getLine();
		}

		if (startLine != endLine) {
			/*
			 * The Eclipse IDE can only use the significant region if it starts and ends on the same line.
			 * Therefore, we here calculate the index of the end of the line.
			 */
			for (int index = offset; index < endOffset; index++) {
				LineAndColumn lineInfo = NodeModelUtils.getLineAndColumn(node, index);
				if (lineInfo.getLine() != startLine) {
					return new TextRegion(offset, index - offset - 1);
				}
			}
		}
		return significantRegion;
	}
```

@lars-reimann  @agonzalezgaliana  could you give it a try
@msujew as you have initially contributed the feature and use it can you also have a look/test

**Comment by lars-reimann** [comment]
Looks good on my end, the test works and the VS Code extension we're developing is not constantly throwing exceptions while typing anymore. 👍

**Comment by agonzalezgaliana** [comment]
Trying to test it also on my side. Since @lars-reimann has already provided positive feedback I'll give feedback only in case the fix does not work.

**Comment by agonzalezgaliana** [comment]
@cdietrich I was able to retest with your proposed fix. Works on my side, no errors appear as before. Thanks!

**Comment by cdietrich** [comment]
@lars-reimann @agonzalezgaliana  thx for testing

**Comment by lars-reimann** [comment]
No worries and thanks for fixing it so quickly.

**Comment by cdietrich** [comment]
fixed in 2.26.0
